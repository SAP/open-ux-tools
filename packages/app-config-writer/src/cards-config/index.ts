import { join, basename } from 'path';
import type { Editor } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import { getPreviewMiddleware, getIntentFromPreviewConfig, getCLIForPreview } from '../common/utils';
import type { MiddlewareConfig as PreviewConfig } from '@sap-ux/preview-middleware';
import type { ToolsLogger } from '@sap-ux/logger';
import { FileName, type Package, readUi5Yaml } from '@sap-ux/project-access';
import { updateMiddlewaresForPreview } from '../common/ui5-yaml';

/**
 * Updates the `ui5.yaml` file to add card generator path to preview middleware configuration.
 *
 * @param {Editor} fs - The `mem-fs-editor` instance used to read and write files.
 * @param {string} basePath - The path to the project root where the `ui5.yaml` file is located.
 * @param {string} [yamlPath] - Path to the `ui5.yaml` configuration file. Defaults to `ui5.yaml` in the project root.
 * @param {ToolsLogger} [logger] - logger instance for logging debug information.
 * @returns {Promise<void>} A promise that resolves when the `ui5.yaml` file has been successfully updated.
 */
async function updateMiddlewareConfigWithGeneratorPath(
    fs: Editor,
    basePath: string,
    yamlPath?: string,
    logger?: ToolsLogger
): Promise<void> {
    const ui5YamlFile = yamlPath ? basename(yamlPath) : FileName.Ui5Yaml;
    const ui5YamlConfig = await readUi5Yaml(basePath, ui5YamlFile, fs);
    const previewMiddleware = await getPreviewMiddleware(ui5YamlConfig);

    if (previewMiddleware) {
        previewMiddleware.configuration ??= {};

        const middlewareConfig = previewMiddleware.configuration as PreviewConfig;
        const cardGeneratorPath = middlewareConfig?.editors?.cardGenerator?.path ?? 'test/flpGeneratorSandbox.html';

        middlewareConfig.editors ??= {};
        middlewareConfig.editors.cardGenerator = {
            path: cardGeneratorPath
        };
        ui5YamlConfig.updateCustomMiddleware(previewMiddleware);
        fs.write(join(basePath, ui5YamlFile), ui5YamlConfig.toString());
        logger?.debug(`Updated preview middleware in ${ui5YamlFile}.`);
    }
}

/**
 * Updates the `package.json` file to include a script for starting the card generator.
 *
 * @param {string} basePath - The path to the project root where the `package.json` file is located.
 * @param {Editor} fs - The `mem-fs-editor` instance used to read and write files.
 * @param {string} [yamlPath] - Optional path to the `ui5.yaml` configuration file for retrieving middleware configurations.
 * @returns {Promise<void>} A promise that resolves when the `package.json` file has been successfully updated.
 */
async function updatePackageJson(basePath: string, fs: Editor, yamlPath?: string): Promise<void> {
    const packageJsonPath = join(basePath, 'package.json');
    if (!fs.exists(packageJsonPath)) {
        throw new Error('package.json not found');
    }

    const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;
    const previewMiddleware = await getPreviewMiddleware(undefined, basePath, yamlPath, fs);
    const intent = getIntentFromPreviewConfig(previewMiddleware?.configuration) ?? '#app-preview';
    const cardGeneratorPath =
        (previewMiddleware?.configuration as PreviewConfig)?.editors?.cardGenerator?.path ??
        'test/flpGeneratorSandbox.html';
    const cliForPreview = await getCLIForPreview(basePath, yamlPath ?? '', fs);

    packageJson.scripts ??= {};
    packageJson.scripts['start-cards-generator'] = `${cliForPreview} --open '${cardGeneratorPath}${intent}'`;
    fs.writeJSON(packageJsonPath, packageJson);
}

/**
 * Enables the card generator configuration for the given application.
 *
 * This function updates the `ui5.yaml` file to include the necessary middlewares for the card generator
 * and modifies the `package.json` file to add a script for starting the card generator.
 *
 * @param {string} basePath - The path to the project root.
 * @param {string} [yamlPath] - Optional path to the `ui5.yaml` configuration file.
 * @param {ToolsLogger} [logger] - Optional logger instance for logging messages.
 * @param {Editor} [fs] - Optional `mem-fs-editor` instance for file system operations. If not provided, a new instance will be created.
 * @returns {Promise<Editor>} A promise that resolves to the updated `mem-fs-editor` instance.
 */
export async function enableCardGeneratorConfig(
    basePath: string,
    yamlPath?: string,
    logger?: ToolsLogger,
    fs?: Editor
): Promise<Editor> {
    fs = fs ?? create(createStorage());
    await updateMiddlewaresForPreview(fs, basePath, yamlPath, logger);
    await updateMiddlewareConfigWithGeneratorPath(fs, basePath, yamlPath, logger);
    await updatePackageJson(basePath, fs, yamlPath);
    return fs;
}
