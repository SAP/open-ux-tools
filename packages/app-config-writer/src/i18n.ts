import type { TOptions } from 'i18next';
import i18next from 'i18next';
import smartLinksRes from './translations/smartlinks-config.json';
import navConfigRes from './translations/navigation-config.json';

export const SMART_LINKS_NS = 'app-config-writer:smartLinksConfig';
export const NAV_CONFIG_NS = 'app-config-writer:navConfig';

/**
 * Initialize i18next with the translations for this module.
 */
export async function initI18n(): Promise<void> {
    await i18next.init({
        resources: {
            en: {
                [SMART_LINKS_NS]: smartLinksRes,
                [NAV_CONFIG_NS]: navConfigRes
            }
        },
        lng: 'en',
        fallbackLng: 'en',
        defaultNS: SMART_LINKS_NS, // Default since first to add translations
        ns: [SMART_LINKS_NS, NAV_CONFIG_NS]
    });
}

/**
 * Helper function facading the call to i18next.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
export function t(key: string, options?: TOptions): string {
    return i18next.t(key, options);
}

initI18n().catch(() => {
    // Ignore any errors since the write will still work
});
