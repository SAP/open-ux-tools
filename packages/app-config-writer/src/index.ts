export { getSmartLinksTargetFromPrompt } from './prompt';
export { generateSmartLinksConfig } from './smartlinks-config';
