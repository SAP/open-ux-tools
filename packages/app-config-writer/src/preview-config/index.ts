import { create, type Editor } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { basename, join } from 'path';
import { FileName, getAllUi5YamlFileNames, getWebappPath, type Package, readUi5Yaml } from '@sap-ux/project-access';
import type { ToolsLogger } from '@sap-ux/logger';
import { prompt, type PromptObject } from 'prompts';
import { createPreviewMiddlewareConfig } from '../variants-config/ui5-yaml';
import type { CustomMiddleware } from '@sap-ux/ui5-config';
import { getPreviewMiddleware, isFioriToolsDeprecatedPreviewConfig } from '../variants-config/utils';
import type { PreviewConfigOptions } from '../types';
import type { FlpConfig, MiddlewareConfig as PreviewConfig } from '@sap-ux/preview-middleware';
import { generateVariantsConfig } from '../variants-config';

const renameMessage = (filename: string): string =>
    `Renamed ${filename} to ${filename.slice(
        0,
        -5
    )}_old.html. This file is no longer needed for the preview. In case there have not been done any modifications you can delete this file. In case of modifications please move the respective content to a custom init script of the preview middleware (see migration information https://www.npmjs.com/package/preview-middleware#migration).`;

/**
 * Converts the local preview files of a project to virtual files.
 *
 * @param basePath - base path to be used for the conversion
 * @param logger logger to report info to the user
 * @param fs - file system reference
 * @returns file system reference
 */
export async function convertToVirtualPreview(basePath: string, logger?: ToolsLogger, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }

    if (!(await checkPrerequisites(basePath, fs, logger))) {
        throw Error('Prerequisites not met. See above log messages for details.');
    }

    if (!(await getExplicitApprovalToAdjustFiles())) {
        logger?.error('Approval not given. Conversion aborted.');
        return fs;
    }

    await updatePreviewMiddlewareConfigs(fs, basePath, logger);
    await renameDefaultSandboxes(fs, basePath);
    await deleteNoLongerUsedFiles(fs, basePath);
    await updateVariantsCreationScript(fs, basePath, logger);

    return fs;
}

/**
 * Update the variants creation script in package.json if needed.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param logger logger to report info to the user
 */
async function updateVariantsCreationScript(fs: Editor, basePath: string, logger?: ToolsLogger): Promise<void> {
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = fs.readJSON(packageJsonPath) as Package | undefined;
    if (packageJson?.scripts?.['start-variants-management']) {
        const ui5Yaml = basename(extractYamlConfigFileName(packageJson?.scripts?.['start-variants-management']));
        const yamlPath = join(basePath, ui5Yaml);
        await generateVariantsConfig(basePath, yamlPath, logger, fs);
    }
}

/**
 * Extract the UI5 yaml configuration file name from the script.
 *
 * @param script - the content of the script from package.json
 * @returns the UI5 yaml configuration file name or 'ui5.yaml' as default
 */
function extractYamlConfigFileName(script: string): string {
    const configParameterValueMatch = / --config (\S*)| -c (\S*)/.exec(script);
    return configParameterValueMatch?.[1] ?? configParameterValueMatch?.[2] ?? FileName.Ui5Yaml;
}

/**
 * Check if the script/-name is valid for the conversion. The script
 * - must contain 'ui5 serve' or 'fiori run' command
 * - must not be a test script
 * - must not relate to 'webapp/index.html'.
 * The script name
 * - must not be 'start-variants-management'
 * - must not be 'start-control-property-editor'.
 *
 * @param scriptName - the name of the script from package.json
 * @param script - the content of the script from package.json
 * @returns indicator if the script is valid
 */
function isValidScript(scriptName: string, script: string | undefined): boolean {
    const isValidScriptName =
        scriptName != 'start-variants-management' && scriptName != 'start-control-property-editor';
    const { path } = extractUrlDetails(script ?? '');
    const isValidScript =
        //eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
        !!(script?.includes('ui5 serve') || script?.includes('fiori run')) &&
        //todo: how to ensure we don't mistake other test scripts for preview scripts?
        !script?.includes('qunit.html') &&
        path != 'index.html';
    return isValidScriptName && isValidScript;
}

/**
 * Update the preview middleware configurations according to the scripts they are being used in package.json.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param logger logger to report info to the user
 */
export async function updatePreviewMiddlewareConfigs(
    fs: Editor,
    basePath: string,
    logger?: ToolsLogger
): Promise<void> {
    const {
        valid: validUi5YamlFileNames,
        invalid: invalidUi5YamlFileNames,
        skipped: skippedUi5YamlFileNames
    } = await getAllUi5YamlFileNames(fs, basePath);
    const unprocessedUi5YamlFileNames = [...validUi5YamlFileNames];
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = fs.readJSON(packageJsonPath) as Package | undefined;
    for (const [scriptName, script] of Object.entries(packageJson?.scripts ?? {})) {
        if (!script || !isValidScript(scriptName, script)) {
            continue;
        }

        const ui5Yaml = basename(extractYamlConfigFileName(script));
        unprocessedUi5YamlFileNames.splice(unprocessedUi5YamlFileNames.indexOf(ui5Yaml), 1);

        if ((invalidUi5YamlFileNames ?? []).includes(ui5Yaml)) {
            logger?.warn(
                `Skipping script ${scriptName} with UI5 yaml configuration file ${ui5Yaml} because it does not comply with the schema.`
            );
            continue;
        }

        if ((skippedUi5YamlFileNames ?? []).includes(ui5Yaml)) {
            logger?.warn(
                `Skipping script ${scriptName} with UI5 yaml configuration file ${ui5Yaml} because the schema validation was not possible for file ${ui5Yaml}.`
            );
            continue;
        }

        if (!validUi5YamlFileNames.includes(ui5Yaml)) {
            logger?.warn(
                `Skipping script ${scriptName} because UI5 yaml configuration file ${ui5Yaml} could not be found.`
            );
            continue;
        }

        await processUi5YamlConfig(fs, basePath, ui5Yaml, script);
        await renameSandbox(fs, basePath, script, logger);
        ensurePreviewMiddlewareDependency(packageJson, fs, packageJsonPath);

        logger?.info(`UI5 yaml configuration file ${ui5Yaml} updated according to script ${scriptName}.`);
    }
    for (const ui5Yaml of unprocessedUi5YamlFileNames) {
        //at least adjust deprecated preview config of unused ui5 yaml configurations
        await processUi5YamlConfig(fs, basePath, ui5Yaml, '', true);
        logger?.warn(
            `UI5 yaml configuration file ${ui5Yaml} it is not being used in any preview script. Outdated preview middleware will be adjusted nevertheless if necessary.`
        );
    }
}

/**
 * Rename the sandbox file used in the given script.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param script - the content of the script
 * @param logger logger to report info to the user
 */
async function renameSandbox(fs: Editor, basePath: string, script: string, logger?: ToolsLogger): Promise<void> {
    const { path: relativePath } = extractUrlDetails(script);
    if (relativePath) {
        const absolutePath = join(await getWebappPath(basePath), relativePath);
        if (fs.exists(absolutePath)) {
            fs.move(absolutePath, absolutePath.replace('.html', '_old.html'));
            logger?.info(renameMessage(relativePath));
        } else {
            logger?.warn(`File ${relativePath} not found. Skipping renaming.`);
        }
    }
}

/**
 * Ensure the @sap/ux-ui5-tooling or @sap-ux/preview-middleware dependency exists in package.json.
 * If not @sap-ux/preview-middleware will be added as devDependency.
 *
 * @param packageJson - the package.json file content
 * @param fs - file system reference
 * @param packageJsonPath - the path to the package.json file
 */
export function ensurePreviewMiddlewareDependency(
    packageJson: Package | undefined,
    fs: Editor,
    packageJsonPath: string
): void {
    if (!packageJson) {
        return;
    }

    const hasDependency = (dependency: string): boolean =>
        !!packageJson?.devDependencies?.[dependency] || !!packageJson?.dependencies?.[dependency];

    const dependencies = ['@sap-ux/preview-middleware', '@sap/ux-ui5-tooling'];
    if (dependencies.some((dependency) => hasDependency(dependency))) {
        return;
    }

    packageJson.devDependencies = { ...packageJson.devDependencies, '@sap-ux/preview-middleware': 'latest' };
    fs.writeJSON(packageJsonPath, packageJson);
}

/**
 * Process the UI5 yaml configuration file.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param ui5Yaml - the name of the UI5 yaml configuration file
 * @param script - the content of the script
 * @param skipPreviewMiddlewareCreation - (default: true) indicator if the preview middleware creation should be skipped if no preview middleware is configured
 */
export async function processUi5YamlConfig(
    fs: Editor,
    basePath: string,
    ui5Yaml: string,
    script: string,
    skipPreviewMiddlewareCreation = false
): Promise<void> {
    const ui5YamlConfig = await readUi5Yaml(basePath, ui5Yaml, fs);
    let previewMiddleware = await getPreviewMiddleware(ui5YamlConfig);

    if (skipPreviewMiddlewareCreation && !previewMiddleware) {
        return;
    }
    if (!previewMiddleware) {
        previewMiddleware = createPreviewMiddlewareConfig(fs, basePath);
    }

    const { path, intent } = extractUrlDetails(script);
    previewMiddleware = updatePreviewMiddlewareConfig(previewMiddleware, intent, path);

    ui5YamlConfig.updateCustomMiddleware(previewMiddleware);
    const yamlPath = join(basePath, ui5Yaml);
    fs.write(yamlPath, ui5YamlConfig.toString());
}

/**
 * Extract the URL details from the script.
 *
 * @param script - the content of the script
 * @returns the URL details
 */
function extractUrlDetails(script: string): {
    path: string | undefined;
    intent: FlpConfig['intent'] | undefined;
} {
    const openParameterValueMatch = / --open ([^"]?\S*)| -o ([^"]?\S*)| --o ([^"]?\S*)/.exec(script);
    let url = openParameterValueMatch?.[1] ?? openParameterValueMatch?.[2] ?? openParameterValueMatch?.[3] ?? undefined;
    url = url?.startsWith('"') ? url.slice(1) : url;
    const path = /^[^?#]+\.html/.exec(url ?? '')?.[0] ?? undefined;
    const intent = /(?<=#)\w+-\w+/.exec(url ?? '')?.[0] ?? undefined;

    return {
        path,
        intent: intent
            ? {
                  object: intent?.split('-')[0],
                  action: intent?.split('-')[1]
              }
            : undefined
    };
}

/**
 * Create a preview middleware configuration.
 *
 * @param previewMiddleware - the preview middleware configuration
 * @param intent - the intent
 * @param path - the flp path
 * @returns the preview middleware configuration
 */
export function updatePreviewMiddlewareConfig(
    previewMiddleware: CustomMiddleware<PreviewConfigOptions>,
    intent: FlpConfig['intent'] | undefined,
    path: string | undefined
): CustomMiddleware<PreviewConfigOptions> {
    const newMiddlewareConfig = sanitizePreviewMiddleware(previewMiddleware);

    const configuration = newMiddlewareConfig.configuration ?? ({} as PreviewConfig);
    configuration.flp = configuration.flp ?? {};

    let writeConfig = false;
    //check path and respect defaults
    if (pathIsFlpPath(path, configuration) && !path?.includes('test/flp.html')) {
        configuration.flp.path = path;
        writeConfig = true;
    }
    //check intent and respect defaults
    if (intent && intent.object != 'app' && intent.action != 'preview') {
        configuration.flp.intent = {
            object: intent.object,
            action: intent.action
        };
        writeConfig = true;
    }

    //eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
    if (writeConfig) {
        newMiddlewareConfig.configuration = configuration;
    }

    return newMiddlewareConfig as CustomMiddleware<PreviewConfig>;
}

/**
 * Check if the path is an FLP path.
 *
 * @param path - the path
 * @param configuration - the preview configuration
 * @returns indicator if the path is an FLP path
 */
function pathIsFlpPath(path: string | undefined, configuration: PreviewConfig): boolean {
    return (
        !!path &&
        (configuration.rta?.editors?.filter((editor) => editor.path === path)?.length === 0 || true) &&
        (configuration.test?.filter((test) => test.path === path)?.length === 0 || true)
    );
}

/**
 * Sanitize the preview middleware. In case of an outdated preview configuration the following changes will be applied:
 * - property 'ui5Theme' will be moved to 'flp.theme'.
 * - no longer used property 'component' will be removed.
 *
 * @param previewMiddleware - the preview middleware
 * @returns the sanitized preview middleware
 */
export function sanitizePreviewMiddleware(
    previewMiddleware: CustomMiddleware<PreviewConfigOptions>
): CustomMiddleware<PreviewConfig | undefined> {
    if (!isFioriToolsDeprecatedPreviewConfig(previewMiddleware.configuration)) {
        return previewMiddleware as CustomMiddleware<PreviewConfig>;
    }
    const ui5Theme = previewMiddleware.configuration.ui5Theme;
    delete (previewMiddleware as CustomMiddleware<PreviewConfig | undefined>).configuration;

    if (!ui5Theme) {
        return previewMiddleware as unknown as CustomMiddleware<undefined>;
    }

    const configuration = {} as PreviewConfig;
    configuration.flp = {};
    configuration.flp.theme = ui5Theme;
    previewMiddleware.configuration = configuration;
    return previewMiddleware as CustomMiddleware<PreviewConfig>;
}

/**
 * Add '_old' to test/flpSandbox.html and test/flpSandboxMockserver.html to indicate that they will no longer be used.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param logger logger to report info to the user
 */
export async function renameDefaultSandboxes(fs: Editor, basePath: string, logger?: ToolsLogger): Promise<void> {
    const defaultSandboxPaths = [join('test', 'flpSandbox.html'), join('test', 'flpSandboxMockserver.html')];
    for (const path of defaultSandboxPaths) {
        const fakeScript = ` --open ${path}`;
        await renameSandbox(fs, basePath, fakeScript, logger);
    }
}

/**
 * Delete files that are no longer used for the virtual preview.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param logger logger to report info to the user
 */
export async function deleteNoLongerUsedFiles(fs: Editor, basePath: string, logger?: ToolsLogger): Promise<void> {
    const webappTestPath = join(await getWebappPath(basePath), 'test');
    [
        join(webappTestPath, 'locate-reuse-libs.js'),
        join(webappTestPath, 'changes_loader.js'),
        join(webappTestPath, 'changes_loader.ts'),
        join(webappTestPath, 'changes_preview.js'),
        join(webappTestPath, 'changes_preview.ts'),
        join(webappTestPath, 'flpSandbox.js'),
        join(webappTestPath, 'flpSandbox.ts'),
        join(webappTestPath, 'initFlpSandbox.js'),
        join(webappTestPath, 'initFlpSandbox.ts')
    ].forEach((path: string): void => {
        if (fs.exists(path)) {
            fs.delete(path);
            logger?.info(
                `Deleted ${join('webapp', 'test', basename(path))}. This file is no longer needed for the preview.`
            );
        }
    });
}

/**
 * Check if the prerequisites for the conversion are met.
 * - UI5 CLI version 3.0.0 or higher is being used.
 * - '@sap/grunt-sapui5-bestpractice-build' is not being used.
 * - '@sap-ux/ui5-middleware-fe-mockserver' or 'cds-plugin-ui5' is being used.
 *
 * @param basePath - base path to be used for the conversion
 * @param fs - file system reference
 * @param logger logger to report info to the user
 * @returns indicator if the prerequisites are met
 */
export async function checkPrerequisites(basePath: string, fs: Editor, logger?: ToolsLogger): Promise<boolean> {
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = fs.readJSON(packageJsonPath) as Package | undefined;
    let prerequisitesMet = true;

    if (!packageJson) {
        throw Error(`File 'package.json' not found at ${basePath}`);
    }

    const sapui5BestpracticeBuildExists =
        !!packageJson?.devDependencies?.['@sap/grunt-sapui5-bestpractice-build'] ||
        !!packageJson?.dependencies?.['@sap/grunt-sapui5-bestpractice-build'];
    if (sapui5BestpracticeBuildExists) {
        logger?.error(
            "A conversion from '@sap/grunt-sapui5-bestpractice-build' is not supported. Please migrate to UI5 CLI version 3.0.0 or higher first. See https://sap.github.io/ui5-tooling/v3/updates/migrate-v3/ for more information."
        );
        prerequisitesMet = false;
    }

    const ui5CliVersion = packageJson?.devDependencies?.['@ui5/cli'] ?? packageJson?.dependencies?.['@ui5/cli'] ?? '0';
    if (parseInt(ui5CliVersion.split('.')[0], 10) < 3) {
        logger?.error(
            'UI5 CLI version 3.0.0 or higher is required to convert the preview to virtual files. See https://sap.github.io/ui5-tooling/v3/updates/migrate-v3/ for more information.'
        );
        prerequisitesMet = false;
    }

    const ui5MiddlewareMockserverExists =
        !!packageJson?.devDependencies?.['@sap-ux/ui5-middleware-fe-mockserver'] ||
        !!packageJson?.dependencies?.['@sap-ux/ui5-middleware-fe-mockserver'];
    const cdsPluginUi5Exists =
        !!packageJson?.devDependencies?.['cds-plugin-ui5'] || !!packageJson?.dependencies?.['cds-plugin-ui5'];
    if (!ui5MiddlewareMockserverExists && !cdsPluginUi5Exists) {
        logger?.error(
            "A conversion from 'sap/ui/core/util/MockServer' is not supported. Please migrate to '@sap-ux/ui5-middleware-fe-mockserver' first (details see https://www.npmjs.com/package/@sap-ux/ui5-middleware-fe-mockserver)."
        );
        prerequisitesMet = false;
    }

    return prerequisitesMet;
}

/**
 * Get the explicit approval form the user to do the conversion.
 *
 * @returns Explicit user approval to do the conversion.
 */
async function getExplicitApprovalToAdjustFiles(): Promise<boolean> {
    const question: PromptObject = {
        name: 'approval',
        type: 'confirm',
        initial: false,
        message:
            'The converter will rename html files and delete js/ts files used for the existing preview and configure the usage of virtual files instead. Do you want to proceed with the conversion?'
    };

    return !!(await prompt([question])).approval;
}
