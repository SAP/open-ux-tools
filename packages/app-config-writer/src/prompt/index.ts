export { getSmartLinksTargetFromPrompt, promptUserPass } from './smartlinks-config';
export { promptInboundNavigationConfig } from './navigation-config';
