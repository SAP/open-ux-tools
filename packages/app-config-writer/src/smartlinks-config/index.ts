export * from './generateSmartLinks';
export { getTargetDefinition, getLocalStoredCredentials } from './utils';
