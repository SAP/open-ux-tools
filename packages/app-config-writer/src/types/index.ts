export * from './smartLinks';
export * from './variantsConfig';
