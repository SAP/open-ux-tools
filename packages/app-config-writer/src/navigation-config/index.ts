import type { Manifest, ManifestNamespace } from '@sap-ux/project-access';
import { FileName, getWebappPath } from '@sap-ux/project-access';
import type { Editor } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import { mergeObjects } from '@sap-ux/ui5-config';
import { join } from 'path';
import { NAV_CONFIG_NS, t } from '../i18n';

/**
 * Adds a basic inbound navigation configuration to the application manifest.
 *
 * @param appRootPath app path
 * @param inboundConfig the inbound configuration to be written
 * @param inboundConfig.semanticObject semantic object
 * @param inboundConfig.action action
 * @param inboundConfig.title Represents a title; to make this property language dependent (recommended), use a key in double curly brackets '{{key}}' and add to i18n file
 * @param inboundConfig.subTitle optional, Represents a subtitle; to make this property language dependent (recommended), use a key in double curly brackets '{{key}}' and add to i18n file
 * @param overwrite overwrite existing config
 * @param fs file system reference
 * @returns file system reference
 */
export async function generateInboundNavigationConfig(
    appRootPath: string,
    { semanticObject, action, title, subTitle }: ManifestNamespace.Inbound[string],
    overwrite = false,
    fs?: Editor
): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    const { manifest, manifestPath } = await readManifest(appRootPath, fs);
    const inboundKey = `${semanticObject}-${action}`;

    if (!overwrite && manifest['sap.app'].crossNavigation?.inbounds[inboundKey]) {
        throw Error(t('error.inboundExists', { inboundKey, ns: NAV_CONFIG_NS }));
    }

    const inbound = {
        [inboundKey]: {
            semanticObject,
            action,
            title,
            subTitle,
            signature: {
                parameters: {},
                additionalParameters: 'allowed'
            }
        }
    };
    const crossNavigation: Manifest['sap.app']['crossNavigation'] = Object.assign(
        manifest['sap.app'].crossNavigation ?? {},
        {
            inbounds: mergeObjects(manifest['sap.app'].crossNavigation?.inbounds, inbound)
        }
    );

    fs.extendJSON(manifestPath, { 'sap.app': Object.assign(manifest['sap.app'], { crossNavigation }) });
    return fs;
}

/**
 * Validates the basic manifest structure and existence required for inbound navigation addition.
 *
 * @param appPath path to the application
 * @param fs Editor instance
 * @throws an error specifiying the validation failure
 * @returns the manifest object and manifest path
 */
export async function readManifest(appPath: string, fs: Editor): Promise<{ manifest: Manifest; manifestPath: string }> {
    const manifestPath = join(await getWebappPath(appPath), FileName.Manifest);
    const manifest = fs.readJSON(manifestPath) as unknown as Manifest;

    if (!manifest) {
        throw Error(t('error.manifestNotFound', { path: manifestPath, ns: NAV_CONFIG_NS }));
    }

    if (!manifest['sap.app']) {
        throw Error(t('error.sapAppNotDefined', { ns: NAV_CONFIG_NS }));
    }

    return {
        manifest,
        manifestPath
    };
}
