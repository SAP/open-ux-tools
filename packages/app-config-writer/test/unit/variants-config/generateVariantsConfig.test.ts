import { join } from 'path';
import { generateVariantsConfig } from '../../../src';
import { ToolsLogger } from '@sap-ux/logger';

describe('generateVariantsConfig', () => {
    const basePath = join(__dirname, '../../fixtures/variants-config');

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('add variants configuration to a project', async () => {
        const fs = await generateVariantsConfig(basePath);
        expect(fs.readJSON(join(basePath, 'package.json'))).toMatchSnapshot();
        expect(fs.read(join(basePath, 'ui5.yaml'))).toMatchSnapshot();
    });

    test('adding variants configuration to a non existing project', async () => {
        const errorLogMock = jest.spyOn(ToolsLogger.prototype, 'error').mockImplementation(() => {});
        const basePath = join(__dirname, '../../fixtures/a-folder-that-does-not-exist');
        const fs = await generateVariantsConfig(basePath, new ToolsLogger());
        expect(errorLogMock).toHaveBeenCalledWith(
            `Script 'start-variants-management' cannot be written to package.json. File 'package.json' not found at ${basePath}.`
        );
    });
});
