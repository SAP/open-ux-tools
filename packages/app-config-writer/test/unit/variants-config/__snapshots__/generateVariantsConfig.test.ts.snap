// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateVariantsConfig add variants configuration to a project 1`] = `
Object {
  "name": "test-app",
  "sapux": true,
  "sapuxLayer": "VENDOR",
  "scripts": Object {
    "start-variants-management": "ui5 serve --open \\"/my-variants.html?fiori-tools-rta-mode=true&sap-ui-rta-skip-flex-validation=true&sap-ui-xx-condense-changes=true&sap-ui-xx-viewCache=false#hello-world\\"",
  },
}
`;

exports[`generateVariantsConfig add variants configuration to a project 2`] = `
"specVersion: \\"3.0\\"
metadata:
  name: fe_lrop_v2
type: application
server:
  customMiddleware:
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          intent:
            object: hello
            action: world
        editors:
          rta:
            layer: VENDOR
            endpoints:
              - path: /editor.html
                developerMode: true
              - path: /my-variants.html
"
`;

exports[`generateVariantsConfig add variants configuration to a project with deprecated preview middleware config 1`] = `
Object {
  "description": "An SAP Fiori application.",
  "name": "v2-lrop-0909",
  "sapux": true,
  "sapuxLayer": "VENDOR",
  "scripts": Object {
    "start-variants-management": "fiori run --open \\"/preview.html?fiori-tools-rta-mode=true&sap-ui-rta-skip-flex-validation=true&sap-ui-xx-condense-changes=true&sap-ui-xx-viewCache=false#app-preview\\"",
  },
}
`;

exports[`generateVariantsConfig add variants configuration to a project with deprecated preview middleware config 2`] = `
"specVersion: \\"3.1\\"
metadata:
  name: v2lrop0909
type: application
server:
  customMiddleware:
    - name: fiori-tools-preview
      afterMiddleware: compression
      configuration:
        flp:
          theme: sap_horizon
"
`;
